### PR TITLE
Add ClockPeriod option to limit performance

### PR DESCRIPTION
**Pull request by dobegor**
Add ClockPeriod option which allows end user to limit VM performance by calling Sleep.

### CODE REVIEW

File: vm/core.go [hunk 1]
```diff
@@ -353,6 +357,7 @@ func (i *Instance) Run() (err error) {
 			}
 		}
 		i.insCount++
+		time.Sleep(i.clockPeriod)
```
Comment (db47h) [inline vm/core.go]: This is a systematic function call that will adversely impact performance for all users. Please make it conditional on `i.clockPeriod > 0`.

**Review by db47h — CHANGES_REQUESTED** [review]
Thanks for your PR. Could you please address the requested change?

Also, out of curiosity, why would anyone want to limit the performance of the VM?

**Comment by dobegor** [comment]
As per https://godoc.org/time#Sleep :
“A negative or zero duration causes Sleep to return immediately.”

So I don’t really see a need for that condition as it’s guaranteed by the std lib.

**Comment by dobegor** [comment]
The reason to limit a performance of the VM is simple. I want to add VM to multiplayer game and this prevents from abusing server resources and requires players to obtain more beefy “CPU” to make it run better.

**Comment by db47h** [comment]
Regarding the call to sleep, although it's indeed a no-op if duration is 0, this is still a function call in the hottest code path of the VM, which will slightly impact performance for users not limiting the speed.

Also, if I understand your use case correctly, what you want to achieve is a constant speed per instruction, regardless of the host CPU's speed (for example simulate a 1MHz virtual CPU). Using a sleep will give you speeds that vary greatly since the time it takes to execute the instruction varies from one CPU to the other.

If that's what you want to achieve, I'd gladly help you with the implementation (I just tested a ticker and it gets dreadfully slow, so this needs more digging).

**Comment by dobegor** [comment]
That makes sense. I’ll add the condition. 

Yes, I do know that performance regulation approach wouldn’t be absolutely fair / consistent, but it’s just enough for most similar usecases.

I also forked the VM because we thought that implicit call for opcodes with values other 30 would be confusing for players, and I’ve added floating opcodes (with an unsafe hack to reinterpret stored integers as floats and vice-verse). 

I’d be happy to look into more fair approach to VM CPU performance limitation if you have any suggestions. 

Thank you!

**Comment by db47h** [comment]
Did some additional testing:

- Raw performance on vm/testdata/core.rx => 180MHz
- Add unconditional call to time.Sleep(0) => 129MHz (the function call impact is huge for a no-op)
- with conditional call to sleep(0) => 163MHz (yes, this is a very hot code path, anything added here has an impact)

Now to solving the problem at hand (and if I'm not mistaken on your use case), tickers are just not fast enough for this (see https://gist.github.com/db47h/700d7bf5d230904de4f312c14e85aebd). However, I got something working nicely with time.Sleep:

```go
func (i *Instance) Run() (err error) {

	// ...

	i.insCount = 0
	startTime := time.Now()
	for i.PC < len(i.Mem) {

		// ...

		i.insCount++

		if i.clockPeriod > 0 {
			endTime := time.Now()
			sleepDuration := i.clockPeriod - endTime.Sub(startTime)
			if sleepDuration > 0 {
				time.Sleep(sleepDuration)
			}
			// we may be adding negative times, but this will even out as we run faster instructions.
			startTime = endTime.Add(sleepDuration)
		}
	}
	return nil
}

```

It works for me with i.clockPeriod = time.Second / time.Duration(4e6) (4 MHz) on an AMD FX6300. Anything higher just gets slower due to the overhead of calling sleep. 

Higher top speeds could be achieved by doing sleeps every n-th instruction:

```go
		const modulo = 64
		i.clockPeriod = time.Second / time.Duration(100e6/modulo) // 100 MHz
		if i.clockPeriod > 0 && i.insCount%modulo == 0 {
			// ...
```

Now it all depends on your use case.

**Comment by db47h** [comment]
I had a look at your fork, nice work! Just keep in mind that the implicit calls are part of the VM specification (and a nice performance boost). But since you're getting rid of retro and going your own way, I guess it doesn't matter much. And while you're at it, you might as well redo the whole IN/OUT/WAIT thing  which is horrendous to work with. 

Sorry about the last push to the repository, but I think it might solve your problem nicely. Just have a look at https://github.com/db47h/ngaro/blob/eac932d0a885f64bde8cf7fde4bf2a848a70b4eb/vm/vm.go#L135-L142
and https://github.com/db47h/ngaro/blob/eac932d0a885f64bde8cf7fde4bf2a848a70b4eb/vm/vm.go#L71-L99 
with example usage in: https://github.com/db47h/ngaro/blob/eac932d0a885f64bde8cf7fde4bf2a848a70b4eb/cmd/retro/main.go#L185-L187

In that last bit `period` should be renamed to `frequency` (in KHz).

Tell me what you think.

**Comment by dobegor** [comment]
> But since you're getting rid of retro and going your own way, I guess it doesn't matter much.

I'm on the way of porting the simplest PL/0 compiler. I've no experience at writing interpreters/compilers at all, but I thought we'd need something more user-friendly for in-game computer that retro :) 

> And while you're at it, you might as well redo the whole IN/OUT/WAIT thing which is horrendous to work with.
Yeah, we might replace the ports/IO APIs to work through interrupts once they are implemented.

As for the Ticker option - I find it very convenient to use and much more optimal than plain time.Sleep. Sure it's not absolutely accurate (it's probably impossible to do at all), but it fits nicely in absolutely most cases (especially triggering a game-specific tick to update the in-game computer state). I'll merge it into the fork.

Thanks for implementing this!

**Comment by db47h** [comment]
> I'm on the way of porting the simplest PL/0 compiler. I've no experience at writing interpreters/compilers at all, but I thought we'd need something more user-friendly for in-game computer that retro :)

Yeah, retro and most concatenative languages are something of an acquired taste. Regarding a compiler, [ragel](http://www.colm.net/open-source/ragel/) is IMO the best to implement the lexer (not mentioned on the home page but it can generate Go code, and there are no licensing issues). For the parser, you might be interested in Rob Pike's talk about [Lexical Scanning in Go](https://talks.golang.org/2011/lex.slide#1).

> it's not absolutely accurate

Yes, as it's implemented, the `resolution` option in `ClockLimiter` is only a "max interval" simply because to check how many instructions have been executed between sleeps I do this https://github.com/db47h/ngaro/blob/e81c572fc532a3fc65110e69aa5b58473f9b122f/vm/core.go#L359

instead of using a modulo:

```go
    if i.tickFn != nil && i.insCount%i.tickCount == 0 {
```

which is just slower (yes, I'm that anal about shaving off every clock cycle that I can get). So I cheat a bit in `ClockLimiter` and adjust the requested `resulution` so that the actual tick count between sleeps matches a nice power of 2 to do a fast modulo with a logical AND.

If the precision of the interval between calls to the ticker function is more important than the actual VM speed (in order to sync more accurately with your frame rate for example), you can use periods that are equal to `resulution / some_power_of_2`. Say your game runs at 60fps, that's 16666666 ns per frame (your `resolution`). You can set period to:

- period = 16666666ns / 262144 => the VM will run at about 15.73 MHz
- period = 16666666ns / 524288 => the VM will run at about 31.46 MHz

and so on.... On the other hand, that's a lot of instructions to be executed with no interruption (which user code might be able to exploit), so you could go for a lower `resolution` and get more that one call every frame.

> Thanks for implementing this!

You're very welcome! I'm looking forward to playing a game with an in-game programmable computer :) Please keep me posted!

**Pull request closed without merging**